File: src-tauri/src/commands/git.rs
use crate::models::git::{
    GitBlameLine, GitBranch, GitCommit, GitConfig, GitContributor, GitDiff, GitGraph, GitMergeResult,
    GitRemote, GitRepositoryInfo, GitStash, GitStatus, GitTag,
};
use crate::services::git_engine::GitEngine;
use crate::utils::error::AppError;
use std::sync::Arc;

pub type GitEngineState = Arc<GitEngine>;

// ── 仓库生命周期 ──────────────────────────────────────────────────

#[tauri::command]
pub async fn git_open(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<GitRepositoryInfo, AppError> {
    state.open(&path).await
}

#[tauri::command]
pub async fn git_close(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<(), AppError> {
    state.close(&path).await
}

#[tauri::command]
pub async fn git_is_open(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<bool, AppError> {
    Ok(state.is_open(&path).await)
}

#[tauri::command]
pub async fn git_validate_repo(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<bool, AppError> {
    state.validate_repo(&path).await
}

// ── 状态 & 历史 ───────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_status(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<GitStatus, AppError> {
    state.get_status(&path).await
}

#[tauri::command]
pub async fn git_current_branch(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<String, AppError> {
    state.current_branch(&path).await
}

#[tauri::command]
pub async fn git_get_commits(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    skip: usize,
    limit: usize,
) -> Result<Vec<GitCommit>, AppError> {
    state.get_commits(&path, skip, limit).await
}

#[tauri::command]
pub async fn git_get_commit_detail(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    hash: String,
) -> Result<GitCommit, AppError> {
    state.get_commit_detail(&path, &hash).await
}

// ── 暂存 & 提交 ──────────────────────────────────────────────────

#[tauri::command]
pub async fn git_stage_file(
    state: tauri::State<'_, GitEngineState>,
    repo_path: String,
    file_path: String,
) -> Result<(), AppError> {
    state.stage_file(&repo_path, &file_path).await
}

#[tauri::command]
pub async fn git_unstage_file(
    state: tauri::State<'_, GitEngineState>,
    repo_path: String,
    file_path: String,
) -> Result<(), AppError> {
    state.unstage_file(&repo_path, &file_path).await
}

#[tauri::command]
pub async fn git_stage_all(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<(), AppError> {
    state.stage_all(&path).await
}

#[tauri::command]
pub async fn git_unstage_all(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<(), AppError> {
    state.unstage_all(&path).await
}

#[tauri::command]
pub async fn git_commit(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    message: String,
    author: String,
    email: String,
) -> Result<String, AppError> {
    state.commit(&path, &message, &author, &email).await
}

// ── Diff ──────────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_diff_working(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<GitDiff, AppError> {
    state.get_diff_working(&path).await
}

#[tauri::command]
pub async fn git_get_diff_staged(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<GitDiff, AppError> {
    state.get_diff_staged(&path).await
}

#[tauri::command]
pub async fn git_get_diff_commit(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    hash: String,
) -> Result<GitDiff, AppError> {
    state.get_diff_commit(&path, &hash).await
}

// ── 分支 ──────────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_branches(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<Vec<GitBranch>, AppError> {
    state.get_branches(&path).await
}

#[tauri::command]
pub async fn git_create_branch(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    name: String,
    start_point: Option<String>,
) -> Result<(), AppError> {
    state.create_branch(&path, &name, start_point.as_deref()).await
}

#[tauri::command]
pub async fn git_delete_branch(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    name: String,
) -> Result<(), AppError> {
    state.delete_branch(&path, &name).await
}

#[tauri::command]
pub async fn git_checkout_branch(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    name: String,
) -> Result<(), AppError> {
    state.checkout_branch(&path, &name).await
}

// ── Stash ─────────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_stashes(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<Vec<GitStash>, AppError> {
    state.get_stashes(&path).await
}

#[tauri::command]
pub async fn git_create_stash(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    message: Option<String>,
) -> Result<usize, AppError> {
    state.create_stash(&path, message.as_deref()).await
}

#[tauri::command]
pub async fn git_apply_stash(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    index: usize,
) -> Result<(), AppError> {
    state.apply_stash(&path, index).await
}

#[tauri::command]
pub async fn git_drop_stash(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    index: usize,
) -> Result<(), AppError> {
    state.drop_stash(&path, index).await
}

// ── 分支图 ────────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_graph(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    skip: usize,
    limit: usize,
) -> Result<GitGraph, AppError> {
    state.get_graph(&path, skip, limit).await
}

// ── Remote 操作 ──────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_remotes(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<Vec<GitRemote>, AppError> {
    state.get_remotes(&path).await
}

#[tauri::command]
pub async fn git_push(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    remote: String,
    branch: String,
    force: bool,
) -> Result<String, AppError> {
    state.push(&path, &remote, &branch, force).await
}

#[tauri::command]
pub async fn git_pull(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    remote: String,
    branch: String,
) -> Result<String, AppError> {
    state.pull(&path, &remote, &branch).await
}

#[tauri::command]
pub async fn git_fetch(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    remote: String,
) -> Result<String, AppError> {
    state.fetch(&path, &remote).await
}

// ── Merge & Rebase ──────────────────────────────────────────────

#[tauri::command]
pub async fn git_merge_branch(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    branch_name: String,
) -> Result<GitMergeResult, AppError> {
    state.merge_branch(&path, &branch_name).await
}

#[tauri::command]
pub async fn git_abort_merge(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<(), AppError> {
    state.abort_merge(&path).await
}

#[tauri::command]
pub async fn git_rebase_branch(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    onto_branch: String,
) -> Result<(), AppError> {
    state.rebase_branch(&path, &onto_branch).await
}

#[tauri::command]
pub async fn git_abort_rebase(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<(), AppError> {
    state.abort_rebase(&path).await
}

// ── Tag 操作 ─────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_tags(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<Vec<GitTag>, AppError> {
    state.get_tags(&path).await
}

#[tauri::command]
pub async fn git_create_tag(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    name: String,
    message: Option<String>,
    target: Option<String>,
) -> Result<(), AppError> {
    state.create_tag(&path, &name, message.as_deref(), target.as_deref()).await
}

#[tauri::command]
pub async fn git_delete_tag(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    name: String,
) -> Result<(), AppError> {
    state.delete_tag(&path, &name).await
}

// ── 文件操作 ─────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_discard_file(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    file_path: String,
) -> Result<(), AppError> {
    state.discard_file(&path, &file_path).await
}

#[tauri::command]
pub async fn git_discard_all(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<(), AppError> {
    state.discard_all(&path).await
}

#[tauri::command]
pub async fn git_get_file_content(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    hash: String,
    file_path: String,
) -> Result<String, AppError> {
    state.get_file_content(&path, &hash, &file_path).await
}

// ── Git Config ──────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_config(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<GitConfig, AppError> {
    state.get_config(&path).await
}

// ── Amend 提交 ──────────────────────────────────────────────────

#[tauri::command]
pub async fn git_amend_commit(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    message: String,
    author: String,
    email: String,
) -> Result<String, AppError> {
    state.amend_commit(&path, &message, &author, &email).await
}

// ── Pop Stash ───────────────────────────────────────────────────

#[tauri::command]
pub async fn git_pop_stash(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    index: usize,
) -> Result<(), AppError> {
    state.pop_stash(&path, index).await
}

// ── Cherry-pick ─────────────────────────────────────────────────

#[tauri::command]
pub async fn git_cherry_pick(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    hash: String,
) -> Result<GitMergeResult, AppError> {
    state.cherry_pick(&path, &hash).await
}

// ── 搜索提交 ────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_search_commits(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    query: String,
    field: String,
    skip: usize,
    limit: usize,
) -> Result<Vec<GitCommit>, AppError> {
    state.search_commits(&path, &query, &field, skip, limit).await
}

// ── Blame ───────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_blame_file(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    file_path: String,
) -> Result<Vec<GitBlameLine>, AppError> {
    state.blame_file(&path, &file_path).await
}

// ── 文件历史 ────────────────────────────────────────────────────

#[tauri::command]
pub async fn git_file_history(
    state: tauri::State<'_, GitEngineState>,
    path: String,
    file_path: String,
    skip: usize,
    limit: usize,
) -> Result<Vec<GitCommit>, AppError> {
    state.file_history(&path, &file_path, skip, limit).await
}

// ── 贡献者统计 ─────────────────────────────────────────────────

#[tauri::command]
pub async fn git_get_contributors(
    state: tauri::State<'_, GitEngineState>,
    path: String,
) -> Result<Vec<GitContributor>, AppError> {
    state.get_contributors(&path).await
}

File: src/api/git.ts
import { invokeCommand } from '@/api/base'
import type {
  GitRepositoryInfo,
  GitStatus,
  GitCommit,
  GitDiff,
  GitBranch,
  GitStash,
  GitGraph,
  GitRemote,
  GitTag,
  GitMergeResult,
  GitConfig,
  GitBlameLine,
  GitContributor,
  RebaseEntry,
  RebaseResult,
} from '@/types/git'

// ── 仓库生命周期 ──────────────────────────────────────────────────

export function gitOpen(path: string): Promise<GitRepositoryInfo> {
  return invokeCommand<GitRepositoryInfo>('git_open', { path }, { source: 'GIT' })
}

export function gitClose(path: string): Promise<void> {
  return invokeCommand('git_close', { path }, { source: 'GIT' })
}

export function gitIsOpen(path: string): Promise<boolean> {
  return invokeCommand<boolean>('git_is_open', { path }, { source: 'GIT' })
}

export function gitValidateRepo(path: string): Promise<boolean> {
  return invokeCommand<boolean>('git_validate_repo', { path }, { source: 'GIT' })
}

// ── 状态 & 历史 ───────────────────────────────────────────────────

export function gitGetStatus(path: string): Promise<GitStatus> {
  return invokeCommand<GitStatus>('git_get_status', { path }, { source: 'GIT' })
}

export function gitCurrentBranch(path: string): Promise<string> {
  return invokeCommand<string>('git_current_branch', { path }, { source: 'GIT' })
}

export function gitGetCommits(path: string, skip: number, limit: number): Promise<GitCommit[]> {
  return invokeCommand<GitCommit[]>('git_get_commits', { path, skip, limit }, { source: 'GIT' })
}

export function gitGetCommitDetail(path: string, hash: string): Promise<GitCommit> {
  return invokeCommand<GitCommit>('git_get_commit_detail', { path, hash }, { source: 'GIT' })
}

// ── 暂存 & 提交 ──────────────────────────────────────────────────

export function gitStageFile(repoPath: string, filePath: string): Promise<void> {
  return invokeCommand('git_stage_file', { repoPath, filePath }, { source: 'GIT' })
}

export function gitUnstageFile(repoPath: string, filePath: string): Promise<void> {
  return invokeCommand('git_unstage_file', { repoPath, filePath }, { source: 'GIT' })
}

export function gitStageAll(path: string): Promise<void> {
  return invokeCommand('git_stage_all', { path }, { source: 'GIT' })
}

export function gitUnstageAll(path: string): Promise<void> {
  return invokeCommand('git_unstage_all', { path }, { source: 'GIT' })
}

export function gitCommit(path: string, message: string, author: string, email: string): Promise<string> {
  return invokeCommand<string>('git_commit', { path, message, author, email }, { source: 'GIT' })
}

// ── Diff ──────────────────────────────────────────────────────────

export function gitGetDiffWorking(path: string): Promise<GitDiff> {
  return invokeCommand<GitDiff>('git_get_diff_working', { path }, { source: 'GIT' })
}

export function gitGetDiffStaged(path: string): Promise<GitDiff> {
  return invokeCommand<GitDiff>('git_get_diff_staged', { path }, { source: 'GIT' })
}

export function gitGetDiffCommit(path: string, hash: string): Promise<GitDiff> {
  return invokeCommand<GitDiff>('git_get_diff_commit', { path, hash }, { source: 'GIT' })
}

// ── 分支 ──────────────────────────────────────────────────────────

export function gitGetBranches(path: string): Promise<GitBranch[]> {
  return invokeCommand<GitBranch[]>('git_get_branches', { path }, { source: 'GIT' })
}

export function gitCreateBranch(path: string, name: string, startPoint?: string): Promise<void> {
  return invokeCommand('git_create_branch', { path, name, startPoint }, { source: 'GIT' })
}

export function gitDeleteBranch(path: string, name: string): Promise<void> {
  return invokeCommand('git_delete_branch', { path, name }, { source: 'GIT' })
}

export function gitCheckoutBranch(path: string, name: string): Promise<void> {
  return invokeCommand('git_checkout_branch', { path, name }, { source: 'GIT' })
}

// ── Stash ─────────────────────────────────────────────────────────

export function gitGetStashes(path: string): Promise<GitStash[]> {
  return invokeCommand<GitStash[]>('git_get_stashes', { path }, { source: 'GIT' })
}

export function gitCreateStash(path: string, message?: string): Promise<number> {
  return invokeCommand<number>('git_create_stash', { path, message }, { source: 'GIT' })
}

export function gitApplyStash(path: string, index: number): Promise<void> {
  return invokeCommand('git_apply_stash', { path, index }, { source: 'GIT' })
}

export function gitDropStash(path: string, index: number): Promise<void> {
  return invokeCommand('git_drop_stash', { path, index }, { source: 'GIT' })
}

// ── 分支图 ────────────────────────────────────────────────────────

export function gitGetGraph(path: string, skip: number, limit: number): Promise<GitGraph> {
  return invokeCommand<GitGraph>('git_get_graph', { path, skip, limit }, { source: 'GIT' })
}

// ── Remote 操作 ──────────────────────────────────────────────────

export function gitGetRemotes(path: string): Promise<GitRemote[]> {
  return invokeCommand<GitRemote[]>('git_get_remotes', { path }, { source: 'GIT' })
}

export function gitPush(path: string, remote: string, branch: string, force = false): Promise<string> {
  return invokeCommand<string>('git_push', { path, remote, branch, force }, { source: 'GIT' })
}

export function gitPull(path: string, remote: string, branch: string): Promise<string> {
  return invokeCommand<string>('git_pull', { path, remote, branch }, { source: 'GIT' })
}

export function gitFetch(path: string, remote: string): Promise<string> {
  return invokeCommand<string>('git_fetch', { path, remote }, { source: 'GIT' })
}

// ── Merge & Rebase ──────────────────────────────────────────────

export function gitMergeBranch(path: string, branchName: string): Promise<GitMergeResult> {
  return invokeCommand<GitMergeResult>('git_merge_branch', { path, branchName }, { source: 'GIT' })
}

export function gitAbortMerge(path: string): Promise<void> {
  return invokeCommand('git_abort_merge', { path }, { source: 'GIT' })
}

export function gitRebaseBranch(path: string, ontoBranch: string): Promise<void> {
  return invokeCommand('git_rebase_branch', { path, ontoBranch }, { source: 'GIT' })
}

export function gitAbortRebase(path: string): Promise<void> {
  return invokeCommand('git_abort_rebase', { path }, { source: 'GIT' })
}

// ── Tag 操作 ─────────────────────────────────────────────────────

export function gitGetTags(path: string): Promise<GitTag[]> {
  return invokeCommand<GitTag[]>('git_get_tags', { path }, { source: 'GIT' })
}

export function gitCreateTag(path: string, name: string, message?: string, target?: string): Promise<void> {
  return invokeCommand('git_create_tag', { path, name, message, target }, { source: 'GIT' })
}

export function gitDeleteTag(path: string, name: string): Promise<void> {
  return invokeCommand('git_delete_tag', { path, name }, { source: 'GIT' })
}

// ── 文件操作 ─────────────────────────────────────────────────────

export function gitDiscardFile(path: string, filePath: string): Promise<void> {
  return invokeCommand('git_discard_file', { path, filePath }, { source: 'GIT' })
}

export function gitDiscardAll(path: string): Promise<void> {
  return invokeCommand('git_discard_all', { path }, { source: 'GIT' })
}

export function gitGetFileContent(path: string, hash: string, filePath: string): Promise<string> {
  return invokeCommand<string>('git_get_file_content', { path, hash, filePath }, { source: 'GIT' })
}

// ── Git Config ──────────────────────────────────────────────────

export function gitGetConfig(path: string): Promise<GitConfig> {
  return invokeCommand<GitConfig>('git_get_config', { path }, { source: 'GIT' })
}

// ── Amend 提交 ──────────────────────────────────────────────────

export function gitAmendCommit(path: string, message: string, author: string, email: string): Promise<string> {
  return invokeCommand<string>('git_amend_commit', { path, message, author, email }, { source: 'GIT' })
}

// ── Pop Stash ───────────────────────────────────────────────────

export function gitPopStash(path: string, index: number): Promise<void> {
  return invokeCommand('git_pop_stash', { path, index }, { source: 'GIT' })
}

// ── Cherry-pick ─────────────────────────────────────────────────

export function gitCherryPick(path: string, hash: string): Promise<GitMergeResult> {
  return invokeCommand<GitMergeResult>('git_cherry_pick', { path, hash }, { source: 'GIT' })
}

// ── 搜索提交 ────────────────────────────────────────────────────

export function gitSearchCommits(path: string, query: string, field: string, skip: number, limit: number): Promise<GitCommit[]> {
  return invokeCommand<GitCommit[]>('git_search_commits', { path, query, field, skip, limit }, { source: 'GIT' })
}

// ── Blame ───────────────────────────────────────────────────────

export function gitBlameFile(path: string, filePath: string): Promise<GitBlameLine[]> {
  return invokeCommand<GitBlameLine[]>('git_blame_file', { path, filePath }, { source: 'GIT' })
}

// ── 文件历史 ────────────────────────────────────────────────────

export function gitFileHistory(path: string, filePath: string, skip: number, limit: number): Promise<GitCommit[]> {
  return invokeCommand<GitCommit[]>('git_file_history', { path, filePath, skip, limit }, { source: 'GIT' })
}

// ── 贡献者统计 ──────────────────────────────────────────────────

export function gitGetContributors(path: string): Promise<GitContributor[]> {
  return invokeCommand<GitContributor[]>('git_get_contributors', { path }, { source: 'GIT' })
}

// ── 交互式 Rebase ──────────────────────────────────────────────

export function gitInteractiveRebasePlan(path: string, baseCommit: string): Promise<RebaseEntry[]> {
  return invokeCommand<RebaseEntry[]>('git_interactive_rebase_plan', { path, baseCommit }, { source: 'GIT' })
}

export function gitInteractiveRebaseExecute(path: string, baseCommit: string, plan: RebaseEntry[]): Promise<RebaseResult> {
  return invokeCommand<RebaseResult>('git_interactive_rebase_execute', { path, baseCommit, plan }, { source: 'GIT' })
}

export function gitInteractiveRebaseAbort(path: string): Promise<void> {
  return invokeCommand('git_interactive_rebase_abort', { path }, { source: 'GIT' })
}
